### ISSUE TITLE
running test by itself fails

### ISSUE THREAD

**Issue by willkg** [mozilla/antenna#89]
I have this problem where if I run all the tests, they pass, but if I run a single test in isolation, it fails.

```
$ make test
<tests all pass>
$ docker-compose run web bash
app@7ca108504991:/app$ py.test tests/unittest/test_crashstorage.py 
================================================ test session starts ================================================
platform linux -- Python 3.5.2, pytest-2.9.1, py-1.4.31, pluggy-0.3.1
rootdir: /app, inifile: setup.cfg
plugins: catchlog-1.2.2, env-0.6.0, wholenodeid-0.2
collected 1 items 

tests/unittest/test_crashstorage.py F

===================================================== FAILURES ======================================================
_________________________ tests/unittest/test_crashstorage.py::TestCrashStorage::test_flow __________________________
Traceback (most recent call last):
  File "/app/tests/unittest/test_crashstorage.py", line 39, in test_flow
    assert (
AssertionError: assert {'submitted_t...526b42110906'} == {'ProductName'...0+00:00', ...}
  Omitting 3 identical items, use -v to show
  Differing items:
  {'uuid': 'c4db5a17-61d0-436d-aa5c-526b42110906'} != {'uuid': 'de1bb258-cbbf-4589-a673-34f802160918'}
  Right contains more items:
  {'ProductName': 'Test',
   'Version': '1.0',
   'dump_checksums': {'upload_file_minidump': 'e19d5cd5af0378da05f63f891c7467af'}}
  Use -v to get the full diff
----------------------------------------------- Captured stderr call ------------------------------------------------
[2011-09-06 00:00:00 +0000] [INFO] antenna.app: c4db5a17-61d0-436d-aa5c-526b42110906 received
[2011-09-06 00:00:00 +0000] [INFO] antenna.app: c4db5a17-61d0-436d-aa5c-526b42110906 accepted
[2011-09-06 00:00:00 +0000] [INFO] antenna.external.crashstorage_base: crash no-op: c4db5a17-61d0-436d-aa5c-526b42110906 [('submitted_timestamp', '2011-09-06T00:00:00+00:00'), ('timestamp', 1315267200.0), ('type_tag', 'bp'), ('uuid', 'c4db5a17-61d0-436d-aa5c-526b42110906')] []
[2011-09-06 00:00:00 +0000] [INFO] antenna.app: c4db5a17-61d0-436d-aa5c-526b42110906 saved
============================================= 1 failed in 0.79 seconds ==============================================
app@7ca108504991:/app$ 
```

This issue covers looking into this further because it's annoying and it might suggest there's something wrong in the test scaffolding or code.

**Comment by willkg** [comment]
There are a few issues here. First is that the expected crash id is different. This test sets `uuid` in the reported crash, so the crash id should be set as the `uuid` value. In this way, the test should be stable and always involve the same crash id. Further, the left side (which is computed) is missing the metadata for the crash.

I instrumented `.on_post()` and poked around and something is wrong with either the HTTP POST request or extracting the POST data because the `raw_crash` coming out of `.extract_payload()` is an empty dict. It's empty because `FieldStorage()` has an empty list after consuming the data.

This only happens if we run the test in isolation.

Thoughts:

1. Is py.test doing some setup when running all the tests that it doesn't do when running one test?
2. Is the app or the test scaffolding setting up some global state when running all the tests that doesn't happen when running one test?
3. Why does this affect HTTP POST body extraction?
4. Has this always been a problem or was it introduced with a specific commit?

**Comment by willkg** [comment]
I found a weird discrepancy which is likely the problem and also suggests this is a problem with the test harness and not a problem with the app.

If I run all the tests, then when the `tests/unittest/test_crashstore.py` test runs, then in `.on_post()`, `req.stream` is a `wsgiref.validate.InputWrapper` instance.

If I run just `tests/unittest/test_crashstore.py`, then in `.on_post()`, `req.stream` is a `falcon.request.request_helpers.Body` instance.

That's not right--it should always be a `wsgiref.validate.InputWrapper` instance.

I traced it down to Falcon. It has a global variable where it maybe wraps the request stream in a `falcon.request.request_helpers.Body` in certain cases. One of the tests verifies that `.extract_payload()` works and this test passes in a different kind of stream which flips that global to `False`. Once it's flipped to `False` all the tests work fine. If that is `True`, then the tests break. I'm pretty sure the problem is in `FieldStorage` where it doesn't work with `falcon.request.request_helpers.Body` in some way.

I didn't look into the FieldStorage thing, but that's probably something I should do.

Instead what I did was to fix the test scaffolding so that every time we build an app, we set that global to `False`. I included a comment explaining the situation.
